### PR TITLE
AlxVeml6040 - Basic Implementation

### PR DESCRIPTION
**Pull request by gregorhudin**

### CODE REVIEW

File: Ext/alxVeml6040.h [hunk 1]
```diff
@@ -116,9 +116,9 @@ typedef union
 		AlxVeml6040_RegEnum_0x00_ColorSensorEnable SD : 1;
 		AlxVeml6040_RegEnum_0x00_ModeSelect AF : 1;
 		AlxVeml6040_RegEnum_0x00_TriggerEnable TRIG : 1;
-		uint8_t unused_3 : 1;  // bit 3 is unused
```
Comment (tomazvidovic) [inline Ext/alxVeml6040.h]: Old system, I think we do not need this comment, it is clear enough from variable name..

File: Ext/alxVeml6040.h [hunk 2]
```diff
@@ -180,11 +180,6 @@ typedef union
 } AlxVeml6040_RegVal_0x0B_W_DATA;
 
 
-//******************************************************************************
```
Comment (tomazvidovic) [inline Ext/alxVeml6040.h]: It is better to just delete all empty sections

File: Ext/alxVeml6040.h [hunk 3]
```diff
@@ -245,10 +235,8 @@ typedef struct
 //******************************************************************************
 typedef struct
 {
-	// Objects - External
```
Comment (tomazvidovic) [inline Ext/alxVeml6040.h]: Old system, currently we just have:
- Const
- Parameters
- Variables
- Info

File: Ext/alxVeml6040.c [hunk 4]
```diff
@@ -114,16 +112,16 @@ Alx_Status AlxVeml6040_Init(AlxVeml6040* me)
 	ALX_VEML6040_ASSERT(me->wasCtorCalled == true);
 	ALX_VEML6040_ASSERT(me->isInit == false);
 
-	// Prepare variables
```
Comment (tomazvidovic) [inline Ext/alxVeml6040.c]: Old system, we will now use always "Local variables"

File: Ext/alxVeml6040.c [hunk 5]
```diff
@@ -114,16 +112,16 @@ Alx_Status AlxVeml6040_Init(AlxVeml6040* me)
...
+	// Local variables
 	Alx_Status status = Alx_Err;
 
 	// Init I2C
 	status = AlxI2c_Init(me->i2c);
-	if (status != Alx_Ok) { ALX_VEML6040_TRACE("Err_AlxI2c_Init"); return status; }
```
Comment (tomazvidovic) [inline Ext/alxVeml6040.c]: Old system, here is OK, to just write "Err" everywhere for TRACE, if error happens, filename, function name & line number will be traced via UART, you can quickly find by this where this happened

File: Ext/alxVeml6040.c [hunk 6]
```diff
@@ -114,16 +112,16 @@ Alx_Status AlxVeml6040_Init(AlxVeml6040* me)
...
+	if (status != Alx_Ok) { ALX_VEML6040_TRACE("Err"); return status; }
 
 	// Check if slave ready
-	status = AlxI2c_Master_IsSlaveReady(me->i2c, me->i2cAddr, 3, 1000);
```
Comment (tomazvidovic) [inline Ext/alxVeml6040.c]: here it should be me->i2cNumOfTries, me->i2cTimeout_ms ...

File: Ext/alxVeml6040.c [hunk 7]
```diff
@@ -133,12 +131,12 @@ Alx_Status AlxVeml6040_Init(AlxVeml6040* me)
...
 	// Write registers
 	status = AlxVeml6040_Reg_Write_All(me);
-	if (status != Alx_Ok) { ALX_VEML6040_TRACE("Err_Reg_WriteNonClkVal"); return status;}
+	if (status != Alx_Ok) { ALX_VEML6040_TRACE("Err"); return status; }
 
 	// Set isInit
 	me->isInit = true;
 
-	// Return OK
+	// Return
```
Comment (tomazvidovic) [inline Ext/alxVeml6040.c:139]: new system, we will just use Return it is more clear

File: Ext/alxVeml6040.c [hunk 8]
```diff
@@ -181,15 +181,17 @@ Alx_Status AlxVeml6040_GetR_raw(AlxVeml6040* me, uint16_t* R_raw)
 	ALX_VEML6040_ASSERT(me->wasCtorCalled == true);
 	ALX_VEML6040_ASSERT(me->isInit == true);
 
-	// Prepare variables
+	// Local variables
 	Alx_Status status = Alx_Err;
 
-	// Read R color data
+	// Read
```
Comment (tomazvidovic) [inline Ext/alxVeml6040.c:187]: Here I found out that comments that indicate code steps its is better that they are short..
So here I just Read

File: Ext/alxVeml6040.c [hunk 9]
```diff
@@ -181,15 +181,17 @@ Alx_Status AlxVeml6040_GetR_raw(AlxVeml6040* me, uint16_t* R_raw)
...
 	status = AlxVeml6040_Reg_Read(me, &me->reg._0x08_R_DATA);
-	if (status != Alx_Ok) { ALX_VEML6040_TRACE("Err_AlxVeml6040_GetR_raw"); return status; }
+	if (status != Alx_Ok) { ALX_VEML6040_TRACE("Err"); return status; }
 
-	// Return OK and data
+	// Set
```
Comment (tomazvidovic) [inline Ext/alxVeml6040.c:191]: Also short comment just Set

File: Ext/alxVeml6040.c [hunk 10]
```diff
@@ -154,17 +152,19 @@ Alx_Status AlxVeml6040_DeInit(AlxVeml6040* me)
 	ALX_VEML6040_ASSERT(me->wasCtorCalled == true);
 	ALX_VEML6040_ASSERT(me->isInit == true);
 
-	// Prepare variables
+	// Local variables
 	Alx_Status status = Alx_Err;
 
+	// TV: TODO Disable IC, so that is stops sampling light sensor
```
Comment (tomazvidovic) [inline Ext/alxVeml6040.c:158]: TODO, here you must disable IC, then DeInit is correct..

**Review by tomazvidovic — CHANGES_REQUESTED** [review]
In general stuff is OK
We must update whole library to new systems, so that all examples are OK...

TODO improve DeInit function

TODO Also when writing WEAK function, use enums always, so you don't need comments..

````
void AlxVeml6040_RegStruct_SetVal(AlxVeml6040* me)
{
	me->reg._0x00_CONF		.val.SD = 0; // enable sensor
	me->reg._0x00_CONF		.val.AF = 0; // Auto mode enabled
	me->reg._0x00_CONF		.val.TRIG = 0; // trigger disabled (if you want use Trigger - set AF = 1)
	me->reg._0x00_CONF		.val.IT = 0b000; // 40ms -> refreshing time for new data
}
````

Like this:

````
void AlxPca9431_RegStruct_SetVal(AlxPca9431* me)
{
	// ADC enable
	me->reg._0Dh_ADC_CONTROL.val.ADC_EN = AdcEn_Enabled;
	me->reg._0Dh_ADC_CONTROL.val.ADC_AVG_EN = AdcAveragingEn_Enabled;
	me->reg._0Dh_ADC_CONTROL.val.ADC_RATE = AdcRate_1ShotConversion;
	me->reg._21h_VOUTLDO_OCP.val.VOUTLDO_OCP = VOutLdoSetOCP_70mA;	// 0x08->40mA, 0x0E->70mA, 0x12->90mA, 0x1C->140mA

	// ADC sample enable
	me->reg._0Eh_Sample_EN.val.NTC_ADC_EN   = NtcAdcSamplingEn_Disabled;
	me->reg._0Eh_Sample_EN.val.OTP_ADC_EN   = OtpAdcSamplingEn_Disabled;
	me->reg._0Eh_Sample_EN.val.IOUT_ADC_EN  = IOutAdcSamplingEn_Enabled;
	me->reg._0Eh_Sample_EN.val.VOUT_ADC_EN  = VOutAdcSamplingEn_Disabled;
	me->reg._0Eh_Sample_EN.val.IRECT_ADC_EN = IRectAdcSamplingEn_Disabled;
	me->reg._0Eh_Sample_EN.val.VRECT_ADC_EN = VRectAdcSamplingEn_Enabled;
	me->reg._0Eh_Sample_EN.val.VTUNE_ADC_EN = VTuneAdcSamplingEn_Disabled;
}
````


Connected to 
auralix/alx-202-af-23-auralix-c-cpp-lib-wiki#39

File: Ext/alxVeml6040.c (hunk 10 above)
Comment (gregorhudin) [reply Ext/alxVeml6040.c:158]: // Veml6040 Set To Default - Disabled IC, so that is stops sampling light sensor
	AlxVeml6040_RegStruct_SetToDefault(me);

**Comment by gregorhudin** [comment]
me->reg._0x00_CONF		.val.SD		= ColorSensor_Enabled;
	me->reg._0x00_CONF		.val.AF		= AutoMode;
	me->reg._0x00_CONF		.val.TRIG	        = TriggerDisabled;
	me->reg._0x00_CONF		.val.IT		= IntegrationTimeSelect_40_ms;

File: Ext/alxVeml6040.c [hunk 11]
```diff
@@ -155,7 +155,8 @@ Alx_Status AlxVeml6040_DeInit(AlxVeml6040* me)
 	// Local variables
 	Alx_Status status = Alx_Err;
 
-	// TV: TODO Disable IC, so that is stops sampling light sensor
+	// Disabled IC
+	AlxVeml6040_RegStruct_SetToDefault(me);
```
Comment (tomazvidovic) [inline Ext/alxVeml6040.c:159]: You forgot to call regWrite to set actual IC..

Comment (gregorhudin) [reply Ext/alxVeml6040.c:159]: // Set register struct values to default
	AlxVeml6040_RegStruct_SetToDefault(me);

	// Write registers - Disabled IC
	status = AlxVeml6040_Reg_Write_All(me);
	if (status != Alx_Ok) { ALX_VEML6040_TRACE("Err"); return status; }

**Review by tomazvidovic — APPROVED** [review]

**Pull request merged**